include/deal.II/differentiation/ad/ad_drivers.h

```diff
@@ -92,56 +92,46 @@ namespace Differentiation
 
 
     /**
-     * A collection of types used within the context of auto-differentiable
-     * numbers.
+     * A struct defining a collection of types used within the context of
+     * auto-differentiable numbers.
      */
-    namespace types
+    template <typename ADNumberType, typename T = void>
+    struct Types
     {
       /**
-       * Typedef for tape indices. ADOL-C uses short integers, so
-       * we restrict ourselves to similar types.
+       * Typedef for tape indices.
        */
-      using tape_index = unsigned short;
+      using tape_index = unsigned int;
 
       /**
        * Typedef for tape buffer sizes.
        */
       using tape_buffer_sizes = unsigned int;
-    } // namespace types
+    }; // struct Types
 
 
     /**
-     * A collection of special numbers used within the context of
-     * auto-differentiable numbers.
+     * A struct defining a collection of special numbers used within the
+     * context of auto-differentiable numbers.
      */
-    namespace numbers
+    template <typename ADNumberType, typename T = void>
+    struct Numbers
     {
       /**
        * A tape index that is unusable and can be used to invalidate recording
        * operations.
-       *
-       * @note ADOL-C doesn't allow us to record to this reserved tape (i.e. can't
-       * write it to file), so we can safely use it as an invalidation case. In
-       * general, we want the user to be able to record to a tape if they'd
-       * like.
        */
-      static const types::tape_index invalid_tape_index = 0;
+      // Note: We stipulate in the documentation for the helper classes that the
+      // valid tape index range is (invalid_tape_index,max_tape_index).
+      static const typename Types<ADNumberType>::tape_index invalid_tape_index =
+        0u;
 
       /**
        * The maximum number of tapes that can be written on one process.
        */
-#ifdef DEAL_II_WITH_ADOLC
-      // Note: This value is a limitation of ADOL-C, and not something that we
-      // have control over. See test adolc/helper_tape_index_01.cc for
-      // verification that we cannot use or exceed this value. This value is
-      // defined as TBUFNUM; see
-      // https://gitlab.com/adol-c/adol-c/blob/master/ADOL-C/include/adolc/internal/usrparms.h#L34
-      static const types::tape_index max_tape_index = TBUFNUM;
-#else
-      static const types::tape_index max_tape_index =
-        std::numeric_limits<types::tape_index>::max();
-#endif // DEAL_II_WITH_ADOLC
-    }  // namespace numbers
+      static const typename Types<ADNumberType>::tape_index max_tape_index =
+        std::numeric_limits<typename Types<ADNumberType>::tape_index>::max();
```

Comment:
that's the maximal value you can represent in this type. but is it also the maximal value adol-c understands? I thought they are using `short`.

Reply:
This is the non-specialised class that you've comment on. The specialisation contains this definition:
```c++
      /**
       * The maximum number of tapes that can be written on one process.
       */
      // Note: This value is a limitation of ADOL-C, and not something that we
      // have control over. See test adolc/helper_tape_index_01.cc for
      // verification that we cannot use or exceed this value. This value is
      // defined as TBUFNUM; see
      // https://gitlab.com/adol-c/adol-c/blob/master/ADOL-C/include/adolc/internal/usrparms.h#L34
      static const typename Types<ADNumberType>::tape_index max_tape_index =
        TBUFNUM;
```

Reply:
Ah, good point!